## Numpy is a powerfull python library working with numbers
## This is powerfull as well as memory efficient 
## This comes preinstalled with anaconda but if do not have anaconda
## With command prompt it can install by command "pip install numpy"

In [1]:
import numpy as np

### Creating numpy array

In [2]:
#Numpy array
a = np.array([1,2,3,4,5,6,7])
#Simple array
b=[1,2,3,4,5,6,7]

In [3]:
a.itemsize

8

In [5]:
#total size of numpy array
a.itemsize*a.size

56

In [6]:
import sys
sys.getsizeof(b)

120

### Here is th diference 
   * Numpy array size is 56 while simple array size is 120
   * Numpy array can be smaller

In [7]:
a = np.array([1,2,3,4,5,6,7],dtype="int8")

In [8]:
a.itemsize*a.size

7

In [9]:
# ordinary array are made up of int32
a = np.array([1,2,3,4,5,6,7])

In [10]:
type(a)

numpy.ndarray

In [13]:
a.dtype

dtype('int64')

### So every character take 8 bite of space and hence total size is 7 * 8 = 56

### So for efficient program always choose dtype as per your requirement

In [14]:
a = np.array([1,2,3,4,5,6,7],dtype="int32")

In [16]:
np.sort(a)

array([1, 2, 3, 4, 5, 6, 7], dtype=int32)

In [17]:
a

array([1, 2, 3, 4, 5, 6, 7], dtype=int32)

In [21]:
np.delete(a,6)

array([1, 2, 3, 4, 5, 6], dtype=int32)

In [23]:
a

array([1, 2, 3, 4, 5, 6, 7], dtype=int32)

In [24]:
a=np.delete(a,6)

### Reshaping of array 
 ##### In this we can transform a array into another only condion is the number of element should be match before and after

In [25]:
a.reshape((3,2))

array([[1, 2],
       [3, 4],
       [5, 6]], dtype=int32)

In [26]:
a.reshape((6,))

array([1, 2, 3, 4, 5, 6], dtype=int32)

### Concatinatio of array 
   * adding one element
   * adding more than one element

In [27]:
np.concatenate((a,[5]))

array([1, 2, 3, 4, 5, 6, 5])

In [28]:
np.concatenate((a,[8]))

array([1, 2, 3, 4, 5, 6, 8])

In [29]:
a = np.concatenate((a,[7,8]))

In [30]:
a.reshape((4,2))

array([[1, 2],
       [3, 4],
       [5, 6],
       [7, 8]])

In [31]:
arr=np.zeros((5,5))

In [88]:
arr

array([[1, 1, 1, 1, 1],
       [1, 0, 0, 0, 1],
       [1, 0, 9, 0, 1],
       [1, 0, 0, 0, 1],
       [1, 1, 1, 1, 1]], dtype=int32)

In [32]:
arr=np.ones((5,5))

In [89]:
arr

array([[1, 1, 1, 1, 1],
       [1, 0, 0, 0, 1],
       [1, 0, 9, 0, 1],
       [1, 0, 0, 0, 1],
       [1, 1, 1, 1, 1]], dtype=int32)

In [33]:
temp=np.zeros((3,3))

In [34]:
temp[1,1]=9

In [35]:
arr[1:-1,1:-1]=temp

In [36]:
arr

array([[1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 1.],
       [1., 0., 9., 0., 1.],
       [1., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1.]])

In [37]:
arr=arr.astype("int32")

In [38]:
arr

array([[1, 1, 1, 1, 1],
       [1, 0, 0, 0, 1],
       [1, 0, 9, 0, 1],
       [1, 0, 0, 0, 1],
       [1, 1, 1, 1, 1]], dtype=int32)

### Equaly distributed values in a range i.e all elemets are in lenear relation ship

In [46]:
ar=np.linspace(1,10,num=10,dtype="int16")
ar

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=int16)

### Numpy random values 

In [47]:
rand=np.random.rand(3,2)

In [48]:
rand

array([[0.56653886, 0.42453774],
       [0.82126582, 0.58536402],
       [0.13126072, 0.70873706]])

In [52]:
ran=np.random.randint(1,10,size=(4,2))
ran

array([[7, 3],
       [9, 2],
       [5, 5],
       [4, 9]])

In [53]:
ran

array([[7, 3],
       [9, 2],
       [5, 5],
       [4, 9]])

### Lenear algebra things 

In [54]:
m=np.identity(4)

In [55]:
m

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [57]:
np.linalg.det(m)

1.0

In [58]:
f=np.full((4,4),9,dtype="int16")

In [59]:
f

array([[9, 9, 9, 9],
       [9, 9, 9, 9],
       [9, 9, 9, 9],
       [9, 9, 9, 9]], dtype=int16)

In [61]:
newf=np.full_like(a,6)

In [62]:
newf

array([6, 6, 6, 6, 6, 6, 6, 6])

In [63]:
np.fill_diagonal(f,4)

In [67]:
f

array([[4, 9, 9, 9],
       [9, 4, 9, 9],
       [9, 9, 4, 9],
       [9, 9, 9, 4]], dtype=int16)

In [68]:
import random as rn

In [69]:
s=np.array([rn.randint(1,8) for i in range(10)])

In [70]:
s

array([6, 5, 8, 8, 1, 1, 2, 3, 3, 7])

In [71]:
np.sort(s)

array([1, 1, 2, 3, 3, 5, 6, 7, 8, 8])

In [72]:
s

array([6, 5, 8, 8, 1, 1, 2, 3, 3, 7])

In [73]:
s.sort

<function ndarray.sort>

In [74]:
s

array([6, 5, 8, 8, 1, 1, 2, 3, 3, 7])

In [75]:
s.sort()

In [76]:
s

array([1, 1, 2, 3, 3, 5, 6, 7, 8, 8])

In [78]:
type(s)

numpy.ndarray

In [79]:
sys.getsizeof(s)

176

In [80]:
s.itemsize

8

In [81]:
s.size

10

In [82]:
s=np.array([1, 1, 2, 3, 3, 5, 6, 7, 8, 8])

In [83]:
sys.getsizeof(s)

176

In [84]:
np.dtype(s[0])

dtype('int64')

In [86]:
s.size*s.itemsize

80

In [87]:
array = [1, 1, 2, 3, 3, 5, 6, 7, 8, 8]
sys.getsizeof(array)

144

# <---------------------------------- THE END ------------------------------>